be/queue.c

```diff
@@ -0,0 +1,443 @@
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+
+/**
+ * @addtogroup be
+ *
+ * @{
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_BE
+#include "lib/trace.h"
+
+#include "be/queue.h"
+
+#include "lib/errno.h"          /* ENOMEM */
+#include "lib/memory.h"         /* M0_ALLOC_ARR */
+#include "lib/buf.h"            /* m0_buf_memcpy */
+#include "lib/misc.h"           /* ergo */
+
+#include "be/op.h"              /* m0_be_op_active */
+
+
+struct be_queue_item {
+	uint64_t         bqi_magic;
+	/** bqq_tl, m0_be_queue::bq_q */
+	struct m0_tlink  bqi_link;
+	char            *bqi_data[];
+};
+
+#define BE_QUEUE_ITEM2BUF(bq, bqi) \
+	M0_BUF_INIT((bq)->bq_cfg.bqc_item_length, &(bqi)->bqi_data)
+
+struct be_queue_wait_op {
+	struct m0_be_op      *bbo_op;
+	struct be_queue_item *bbo_bqi;
+	struct m0_buf         bbo_data;
+	bool                 *bbo_successful;
+	uint64_t              bbo_magic;
+	struct m0_tlink       bbo_link;
+};
+
+M0_TL_DESCR_DEFINE(bqq, "m0_be_queue::bq_q*[]", static,
+		   struct be_queue_item, bqi_link, bqi_magic,
+		   M0_BE_QUEUE_Q_MAGIC, M0_BE_QUEUE_Q_HEAD_MAGIC);
+M0_TL_DEFINE(bqq, static, struct be_queue_item);
+
+M0_TL_DESCR_DEFINE(bqop, "m0_be_queue::bq_op_*[]", static,
+		   struct be_queue_wait_op, bbo_link, bbo_magic,
+		   M0_BE_QUEUE_OP_MAGIC, M0_BE_QUEUE_OP_HEAD_MAGIC);
+M0_TL_DEFINE(bqop, static, struct be_queue_wait_op);
+
+
+static uint64_t be_queue_qitems_nr(struct m0_be_queue *bq)
+{
+	return bq->bq_cfg.bqc_q_size_max + bq->bq_cfg.bqc_producers_nr_max;
+}
+
+static struct be_queue_item *be_queue_qitem(struct m0_be_queue *bq,
+					    uint64_t            index)
+{
+	M0_PRE(index < be_queue_qitems_nr(bq));
+	return (struct be_queue_item *)
+		(bq->bq_qitems + index *
+		 (sizeof(struct be_queue_item) + bq->bq_cfg.bqc_item_length));
+}
+
+static bool be_queue_invariant(struct m0_be_queue *bq)
+{
+	M0_PRE(m0_mutex_is_locked(&bq->bq_lock));
+
+	return bq->bq_enqueued >= bq->bq_dequeued;
+}
+
+M0_INTERNAL int m0_be_queue_init(struct m0_be_queue     *bq,
+                                 struct m0_be_queue_cfg *cfg)
+{
+	uint64_t i;
+
+	M0_ENTRY("bq=%p bqc_q_size_max=%"PRIu64" "
+		 "bqc_producers_nr_max=%"PRIu64" bqc_consumers_nr_max=%"PRIu64,
+		 bq, cfg->bqc_q_size_max,
+		 cfg->bqc_producers_nr_max, cfg->bqc_consumers_nr_max);
+	M0_PRE(M0_IS0(bq));
+	M0_PRE(cfg->bqc_q_size_max > 0);
+	M0_PRE(cfg->bqc_producers_nr_max > 0);
+	M0_PRE(cfg->bqc_consumers_nr_max > 0);
+	M0_PRE(cfg->bqc_item_length > 0);
+	M0_PRE(M0_IS_8ALIGNED(cfg->bqc_item_length));
+
+	bq->bq_cfg = *cfg;
+	bq->bq_the_end = false;
+	bq->bq_enqueued = 0;
+	bq->bq_dequeued = 0;
+	M0_ALLOC_ARR(bq->bq_qitems, be_queue_qitems_nr(bq) *
+		     (sizeof(struct be_queue_item) + cfg->bqc_item_length));
+	M0_ALLOC_ARR(bq->bq_ops_put, bq->bq_cfg.bqc_producers_nr_max);
+	M0_ALLOC_ARR(bq->bq_ops_get, bq->bq_cfg.bqc_consumers_nr_max);
```

Comment:
Array is a region of continuous memory.
Do we really need an array here?
We can separately allocate "q->bq_cfg.bqc_producers_nr_max" number of struct be_queue_wait_op.

Reply:
The arrays were made to simplify the allocations. What you're suggesting would make sense if we have this code to work in kernel, but in user space it doesn't matter.